File: patch_app.py
import json
import re

with open('links.json', 'r', encoding='utf-8') as f:
    links = json.load(f)

updates = {
    1: {"realName": "Greenview at Presidential", "commission": ""},
    2: {"realName": "Bentley Residences in Miami", "commission": "5%"},
    3: {"realName": "Oasis Hallandale Beach", "commission": "6%"},
    4: {"realName": "Viceroy Brickell Residences", "commission": "7%"},
    5: {"realName": "Ora by Casa Tua", "commission": "6%"},
    6: {"realName": "Mercedes-Benz Places Miami", "commission": "6%"},
    7: {"realName": "Lotus Edge Community", "commission": "3%"},
    8: {"realName": "Surf Row Residences", "commission": "5%"},
    9: {"realName": "La Baia North", "commission": "7% plus bonus $25,000"},
    10: {"realName": "THE WELL Bay Harbor Islands", "commission": "6%"},
    11: {"realName": "Glass House Boca Raton", "commission": "3%"},
    12: {"realName": "Bay Harbor Towers", "commission": "7%"},
    13: {"realName": "Andare Residences by Pininfarina", "commission": "5%"},
    14: {"realName": "Sixth&Rio —Fort Lauderdale", "commission": "5%"},
    15: {"realName": "Continuum Club & Residences in North Bay Village", "commission": "6% and $10k bonus"},
    16: {"realName": "Shoma Bay", "commission": "6%"},
    17: {"realName": "Avenia FENDI Casa Residences", "commission": "5%"}
}

with open('app.js', 'r', encoding='utf-8') as f:
    app_str = f.read()

# Instead of parsing the whole JS object, let's just use regex to replace specific properties for each id
for pid in range(1, 18):
    id_str = f"id: {pid},"
    
    # Check if id exists
    if id_str in app_str:
        start_idx = app_str.find(id_str)
        end_idx = app_str.find("},", start_idx)
        if end_idx == -1:
            end_idx = app_str.find("}\n];", start_idx)
            
        block = app_str[start_idx:end_idx]
        
        real_name = updates[pid]["realName"].replace('"', '\\"')
        commission = updates[pid]["commission"].replace('"', '\\"')
        res = json.dumps(links.get(str(pid), []))
        
        # Replace name
        block = re.sub(r'name:\s*".*?",', f'name: "{real_name}",', block)
        # Add commission right after location or name
        if 'commission:' not in block:
            block = re.sub(r'(name:\s*".*?",)', r'\1\n        commission: "'+commission+'",', block)
        
        # Add resources right after images array
        if 'resources:' not in block:
            # find images: [...]
            # this is easier to just append before the end of the block
            # we can just append it to the end of the block
            pass # wait, images is an array, it might span multiple lines.
            
        app_str = app_str[:start_idx] + block + app_str[end_idx:]
        
        # To handle resources cleanly, let's insert it before the closing brace of the object
        # Since end_idx is the closing brace... Wait.
        # Let's just do a regex replace for the whole `images: [...]` to also append resources.
        
with open('app.js', 'w', encoding='utf-8') as f:
    f.write(app_str)

File: patch_app.js
const fs = require('fs');

const links = require('./links.json');

const updates = {
    1: { realName: "Greenview at Presidential", commission: "" },
    2: { realName: "Bentley Residences in Miami", commission: "5%" },
    3: { realName: "Oasis Hallandale Beach", commission: "6%" },
    4: { realName: "Viceroy Brickell Residences", commission: "7%" },
    5: { realName: "Ora by Casa Tua", commission: "6%" },
    6: { realName: "Mercedes-Benz Places Miami", commission: "6%" },
    7: { realName: "Lotus Edge Community", commission: "3%" },
    8: { realName: "Surf Row Residences", commission: "5%" },
    9: { realName: "La Baia North", commission: "7% plus bonus $25,000" },
    10: { realName: "THE WELL Bay Harbor Islands", commission: "6%" },
    11: { realName: "Glass House Boca Raton", commission: "3%" },
    12: { realName: "Bay Harbor Towers", commission: "7%" },
    13: { realName: "Andare Residences by Pininfarina", commission: "5%" },
    14: { realName: "Sixth&Rio —Fort Lauderdale", commission: "5%" },
    15: { realName: "Continuum Club & Residences in North Bay Village", commission: "6% and $10k bonus" },
    16: { realName: "Shoma Bay", commission: "6%" },
    17: { realName: "Avenia FENDI Casa Residences", commission: "5%" }
};

let appStr = fs.readFileSync('app.js', 'utf8');

// We will use eval to parse the projectsData, then stringify and replace.
// But first we must extract just the projectsData string.
const startIdx = appStr.indexOf('const projectsData = [');
const endMarker = '];\n\n// --- APP STATE ---';
const endIdx = appStr.indexOf(endMarker) + 2;

const dataStr = appStr.substring(startIdx + 'const projectsData = '.length, endIdx);

// Evaluate it
let projectsData;
eval('projectsData = ' + dataStr);

// Modify
projectsData.forEach(p => {
    if (updates[p.id]) {
        p.name = updates[p.id].realName;
        p.commission = updates[p.id].commission;
    }
    if (links[p.id]) {
        p.resources = links[p.id];
    } else {
        p.resources = [];
    }
});

// Convert back to JS string
function escapeString(str) {
    if(!str) return '""';
    return '"' + str.replace(/"/g, '\\"').replace(/\n/g, '\\n') + '"';
}

let newJs = '[\n';
projectsData.forEach((p, i) => {
    newJs += '    {\n';
    newJs += `        id: ${p.id},\n`;
    newJs += `        name: ${escapeString(p.name)},\n`;
    newJs += `        location: ${escapeString(p.location)},\n`;
    newJs += `        type: ${escapeString(p.type)},\n`;
    newJs += `        hoa: ${escapeString(p.hoa)},\n`;
    newJs += `        contact: ${escapeString(p.contact)},\n`;
    newJs += `        startingPrice: ${escapeString(p.startingPrice)},\n`;
    newJs += `        size: ${escapeString(p.size)},\n`;
    newJs += `        developer: ${escapeString(p.developer)},\n`;
    newJs += `        architect: ${escapeString(p.architect)},\n`;
    newJs += `        delivery: ${escapeString(p.delivery)},\n`;
    newJs += `        description: \`${p.description}\`,\n`;
    
    newJs += `        commission: ${escapeString(p.commission)},\n`;
    
    newJs += `        amenities: ${JSON.stringify(p.amenities)},\n`;
    newJs += `        images: ${JSON.stringify(p.images, null, 12).replace(/\\n/g, '\n')},\n`;
    
    newJs += `        resources: ${JSON.stringify(p.resources, null, 12).replace(/\\n/g, '\n')}\n`;
    
    newJs += '    }' + (i < projectsData.length - 1 ? ',' : '') + '\n';
});
newJs += ']';

const updatedAppStr = appStr.substring(0, startIdx) + 'const projectsData = ' + newJs + appStr.substring(endIdx);
fs.writeFileSync('app.js', updatedAppStr);
console.log('Successfully patched app.js');
